### ISSUE TITLE
When an 'insertHTML' command should be applied as sibling or child of a styled 'span' element ? 

### ISSUE THREAD

**Issue by javifernandez** [w3c/editing#191]
Lets consider this case:

```
<!doctype html>
<style>
.textBox {
   font: 20px/1 Arial;
   height: 200px;
   border: 2px solid blue;
}
</style>
<script>
    function insertHTML() {
       document.getElementById("theFrame").focus();
       document.execCommand("insertHTML", false, '<a href="www.igalia.com">Igalia</a><div style="display: inline;font-size: large"></div>');
    };
    function onLoad() {
       document.getElementById("theFrame").focus();
       document.execCommand("foreColor", false, "green");
    }
</script>
<body onload="onLoad();">
    <div id="theFrame" class="textBox" contenteditable="true"></div>
    <div>
        <button onclick="insertHTML()">Insert HTML</button>
    </div>
</body>
```

After typing a single character, a new &lt;font&gt; element is inserted as **first child** of the "theFrame" &lt;div&gt; element, which will be the parent of the inline block that will contain all the text. This is the DOM tree generated by any of the major browsers:

```
<div id="theFrame" class=textBox" contenteditable=true>
   <font color="green">XXXXX</font>
</div>
```

Once the insertHTML command is executed, via the "insert HTML" button, the new HTML fragment is added to the DOM tree, and this is where browsers shows an interoperability issue. 

In the case of Safari and Chrome, the new HTML fragment is inserted as **sibling** of the &lt;font&gt; element, producing the following DOM tree:

```
<div id="theFrame" class=textBox" contenteditable=true>
   <font color="green">XXXXX</font>
   <a href="www.igalia.com">Igalia</a>
   <div display="inline; font-size: large"><div>
   YYYY
</div>
```

This behavior implies that the text that follows the new HTML inserted doesn't follow the same style (font color in this case) than the preceding text.

Firefox and Edge insert the new HTML fragment as a **new child** of the &lt;font&gt; element, but they produce a slightly different DOM, since Firefox refuses to insert the empty inline block; this is what Firefox produces after executing the insertHTML command:

```
<div id="theFrame" class=textBox" contenteditable=true>
   <font color="green">
      XXXXX
      <a href="www.igalia.com">Igalia</a>
   </font>
</div>
```

While Edge produces the following DOM after executing the the insertHTML command;

```
<div id="theFrame" class=textBox" contenteditable=true>
   <font color="green">
      XXXXX
     <a href="www.igalia.com">Igalia</a>
     <div display="inline; font-size: large"><div>
   </font>
</div>
```

Additionally, there are still differences between Firefox and Edge regarding where the text following the new HTML fragment is inserted in the DOM tree. 

In the case of Firefox, it's inserted as a **sibling** of the &lt;font&gt; element, producing a DOM tree that looks as follows:

```
<div id="theFrame" class=textBox" contenteditable=true>
   <font color="green">
      XXXXX
      <a href="www.igalia.com">Igalia</a>
   </font>
    YYYY
</div>
```

On the other hand, Edge produces a DOM tree that guarantees the text's **style is preserved** after executing the insertHTML command:

```
<div id="theFrame" class=textBox" contenteditable=true>
   <font color="green">
      XXXXX
     <a href="www.igalia.com">Igalia</a>
     <div display="inline; font-size: large"><div>
     YYYYY
   </font>
</div>
```

**Comment by javifernandez** [comment]
Just to give my opinion, I really think Edge case is the most correct one, both visually and in terms of the DOM tree generated after the execution of the insertHTML command. 

However, it's unfortunate that 3 of the main browsers have the same behavior, at least visually since they generate a different DOM tree, which it's already a source of interoperability issues.

**Comment by johanneswilm** [comment]
I think the sudden appearance of the obsolete `<font>` element [1] is likely one of the first signs to JS devs that contenteditable is unmaintained. I remember running into that around 2012. I don't think any solution that involves HTML 3 elements that have been deprecated since should be encouraged in 2019. Even if that means breaking with the current behavior of all browsers.

[1] https://developer.mozilla.org/en-US/docs/Web/HTML/Element/font

**Comment by javifernandez** [comment]
wow ! got it, thanks.

I'll do some experiments, with your statement in mind, and I'll close the issue if nothing interesting comes out of them.

**Comment by johanneswilm** [comment]
I do think it would make sense to define some behavior involving approved HTML5 syntax. Even though JS editors will define their own behavior, they will often need to diff the DOM and it's quite a bit harder if one needs to take into consideration that the browser creates elements of a type that is discouraged from being used.

**Comment by javifernandez** [comment]
Yes, I understand, and I think it makes sense to define such new behavior based on HTML5 syntax. This issue was mainly related to insertHTML command, but it only appears when used in combination with foreColor, which is the one creating deprecated elements. 

Missing the original text style (font color in this case) is indeed unfortunate, but I guess it should be solved by using a new HTML5 based foreColor command.

**Comment by johanneswilm** [comment]
The place where I encountered the font-element, if I recall correctly, was when hitting enter in the middle of an inline element that was styled in a certain way. This happened on Chrome.
Would it not be possible to make a general rule that the browser should never create pre-HTML5 elements as part of contenteditable and then grep for all instances of font-elements in the source code and replace them with spans? Possibly with inline styles.

**Comment by javifernandez** [comment]
I'm not sure how many backward compatibility issues we would have with such change in the web engine. I guess the simpler solution would be to define a new command that overrides the functionality of the 'foreColor' command. This new command will indeed use only HTML5 elements to apply the style change.

**Comment by johanneswilm** [comment]
Those editors that depend on the creation of the obsolete font-element will likely also crash if you make any other changes to the current browser behavior.
I think we need to decide: either we can fix things, or we cannot. If we can fix things, then pre-HTML5 has to go.

**Comment by javifernandez** [comment]
I'm all for fixing things avoiding pre-HTML5 stuff, but note that this issue is not about the foreColor exec command itself, but about insertHTML. 

The issue here is whether the inserted HTML fragment should be place as child or sibling of the 'font' element. It seems 2 of the main browsers follow each of these behaviors; then, among the ones betting for inserting the HTML as sibling (FF and Edge) there is the interop issue of where to insert the following text; I think continue inserting inline elements as children of the styled element makes more sense so that we can preserve the text style after the insertHTML command.

On the other hand, I think we agree that the ideal solution would be to define new commands that forbids the use of pre-HTML5 and that consider the case I'm describing in this issue. Since none of the main browsers seems to be interoperable, either the spec is not clear about this case or there should be an easy fix, following current spec (even using deprecated elements).

**Comment by johanneswilm** [comment]
Right. But that doesn't make much of a difference. pre-HTML5 elements should not appear under any circumstances - no matter whether the user hits Enter, or you execute one execCommand or the other. That's why I would formulate a general policy that pre-HTML5 content should not be added to contenteditable by the browser under any circumstances. The only thing that makes this case slightly better than it happening upon an Enter-press is that execCommand usage is in itself discouraged and no modern editor uses it any more. 

I also challenge the view that removing the font element breaks existing editors. Using editors on a website generally means that content is being changed and so under almost all circumstances means some kind of setup with server side programming and a database. It's not really possible to have run a page with server side code since before 1997 when HTML 4 was released and not having been hacked since then without updating the site's source code.

So no, I don't think we should be making changes that again involve a font element. If any browser is being changed, it should be changed so that the font element is gone.

**Comment by javifernandez** [comment]
Ok, I've been doing some experimental changes in chrome so that the foreColor command inserts an 'span' with an inline style instead of the deprecated 'font' element. However, this issue still stands, as the subsequent insertHTML command is placed as sibling of such 'span' element, which causes that the following text doesn't have the same style. 

The issue is, is that the expected behavior of the insertHTML command ? Leaving aside that all the major browsers still use the deprecated font element, it seems that there is an interoperability issue regarding whether to insert the new HTML fragment as sibling or child of the styled element (either span or font).

BTW, I'll change the title of the issue to avoid further noise; we can file a new issue to discuss the change of removing completely the use of font and other deprecated pre-HTML5 elements.

**Comment by johanneswilm** [comment]
Ok, so as for the remaining part of the structure - I agree that Edge and Firefox make more sense. Edge probably makes the most sense, but given that Edge will stop being a major browser shortly, and the choice then is between standardizing on the Firefox or the Chrome behavior, I would suggest to go with the Firefox behavior.

**Comment by javifernandez** [comment]
Ok, thanks a lot for the feedback. 

The funny thing is that implementing Firefox behavior in chrome, allowing the new HTML fragment to be inserted inside the previous styled sibling, leads to the behavior Edge is currently implementing, so that the following text is indeed inserted inside the styled element as well. 

I said this is funny because following this path will end up with Firefox being the only one non-interoperable, since the text that follows the new HTML fragment doesn't preserve the style. However, this would be a different issue, more related to the insertText command.

**Comment by johanneswilm** [comment]
I think what is important is that the browsers A) don't produce pre-HTML5 content and B) that they behave similarly C) that what they do makes sense. The prioritization should be in that order (A, B, C). In this case, after A has been solved, B is important, C not so much. It would be good to have Firefox behave like the other two, but given that almost all of execCommand is not used by known editors that run in unpatched browsers, it's not that important what it does, really.

The question about subsequent text is whether the inline node that holds the styling is `inclusive` or not. The right answer to that really depends on what the semantic purpose of the styling is. JS editor frameworks have settings for specific styles to either be inclusive or not inclusive [1][2]. The JS editors know what the semantic purpose of the styling is - the browser do not. For example for the style is normally inclusive for italic and bold, but not for links. A "link" may just be some text that is styled with an underline and with an event listener listening for a click on them to then create a popup within the page. it's really impossible for browsers to tell what is appropriate behavior in any given situation - maybe the green color applied in your example is meant to show that it's a link. Maybe it has a similar function as bold. 

That's why browsers should just focus on doing the same thing rather than try to do the right thing. That way JS editors have an easier time modifying their behavior without having to write special code for each browser engine.


[1] http://prosemirror.net/docs/ref/#model.MarkSpec.inclusive 
[2] https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_utils_bindtwostepcarettoattribute.html#static-function-bindTwoStepCaretToAttribute

**Comment by javifernandez** [comment]
> That's why browsers should just focus on doing the same thing rather than try to do the right thing. That way JS editors have an easier time modifying their behavior without having to write special code for each browser engine.

Yeah, that makes a lot of sense. 
Thanks again.

**Comment by gked** [comment]
I agree with your analysis.
For (A), I'll get some usage data and we can assess if it would be safe to remove generation of <font> element. 

It sounds like consistency in behavior is more important rather than the behavior itself.
The behavior heuristics are subjective and I can see pros/cons for both behaviors.
Ideally, it would be nice to achieve more with less effort which means that FF would make a change to match blink/webkit.

**Comment by masayuki-nakano** [comment]
(well, off topic though, the Firefox behavior is just our (Mozilla's) bug. I'll take a look when I have much time to do it.)

**Comment by javifernandez** [comment]
> (well, off topic though, the Firefox behavior is just our (Mozilla's) bug. I'll take a look when I have much time to do it.)

I'd appreciate if you could paste here a ink to the corresponding bug report, when available.

**Comment by johanneswilm** [comment]
See also [this explanation](https://github.com/ckeditor/ckeditor5/issues/1663#issuecomment-481690751) of how the CKEditor Framework has implemented the different behavior that content typed at the end of links is treated differently than content inserted at the end of bold/italic. 

In CKEditor, just for links, it defined two caret positions - just outside and just inside the styled inline node - whereas usually for inline content there would just be one and the styling would automatically be inherited from the node to the left of the caret. 

This is different from the ProseMirror framework that I referenced earlier, in which there is always just one caret position at the start/end of inline styled nodes, and the schema decides whether a specific type of styling is applied when new inline content is inserted at the right-hand end.

Both frameworks have come up with valid solutions. It shows that it's not really up to browsers to make the final decision of how the DOM should be updated.
